### ISSUE TITLE
Throwing error on invalid JSON string…

### ISSUE THREAD

**Issue by zedrdave** [marshmallow-code/webargs#329]
_This seems too big to be an oversight, so apologies in advance if this is a design choice and it's been covered before (I looked in the issues and didn't see anything):_

When sending an invalid JSON string, eg (note the stray comma):
```
curl -X PUT "http://localhost:5000/bar/0001" -H "accept: application/json" -H "Content-Type: application/json" -d "{ \"foo\": \"bar\" , }"
```

… it seems that parsers are expected to simply return an empty dict `{}`, while silencing any error. Eg `flask_parser`:
```
        # Fail silently so that the webargs parser can handle the error
        if hasattr(req, "get_json"):
            # Flask >= 0.10.x
            json_data = req.get_json(force=force, silent=True)
        else:
            # Flask <= 0.9.x
            json_data = req.json
        if json_data is None:
            return core.missing
```

Despite what the comment implies, I don't think there is any way later on to distinguish between a real missing value, and one resulting from an invalid JSON string…

As expected, assuming an empty dict is compatible with Schema validation options, no error is thrown (and adding an error handler does not seem like it would help).

Am I missing something? Shouldn't JSON parsing errors be caught and trigger a 400 or 422 automatically?

**Comment by sloria** [comment]
Thanks for the report, @zedrdave . I think you're right about not handling JSON correctly. I would also expect an error response to be returned.

I won't have time to look at this in the immediate future. Would you like to send a PR for this? Even a partial PR, e.g. a failing test, would be a big help.

**Comment by sloria** [comment]
After further investigation, all the parsers behave the same way (ignoring invalid JSON and treating it as missing).

I added this to `webargs.testing.CommonTestCase`:

```python
    def test_invalid_json(self, testapp):
        res = testapp.post(
            "/echo",
            '{"foo": "bar", }',
            headers={"Accept": "application/json", "Content-Type": "application/json"},
            expect_errors=True,
        )
        assert res.status_code == 400
```


and it failed for every parser. So it appears this behavior is expected (though perhaps not desired).

I'm still up for changing this, but the behavior should be changed across all parsers.

**Comment by sloria** [comment]
This issue is resolved by #334, which ~~allows `JSONDecodeError` to be raised~~ returns a 400 error response when the client passes an invalid JSON payload.

~~Your code will still need to handle the error appropriately.~~
